File: C/credit.c
#include <cs50.h>
#include <math.h>
#include <stdio.h>

int checksum(long num);
int multiplyAndSum(int lastdig);
int size(long num);

int main(void)
{
    long num = get_long("Whats your number? ");

    if (checksum(num) == 0)
    {
        int firstDigit = num / pow(10, size(num) - 1);
        int firstTwoDigits = num / pow(10, size(num) - 2);
        if ((size(num) == 13 || size(num) == 16) && firstDigit == 4)
            printf("VISA\n");
        else if ((size(num) == 16) && (firstTwoDigits >= 51 && firstTwoDigits <= 55))
            printf("MASTERCARD\n");
        else if ((size(num) == 15) && (firstTwoDigits == 34 || firstTwoDigits == 37))
            printf("AMEX\n");
        else
            printf("INVALID\n");
    }
    else
        printf("INVALID\n");
    return (0);
}

int checksum(long num)
{
    int i = 0;
    int lastDig = 0;
    bool alternate = false;

    while (num > 0)
    {
        if (alternate == true)
        {
            lastDig = num % 10;
            int product = multiplyAndSum(lastDig);
            i = i + product;
        }
        else
        {
            lastDig = num % 10;
            i = i + lastDig;
        }
        alternate = !alternate;
        num = num / 10;
    }
    if (i % 10 == 0)
        return (0);
    return (1);
}

int multiplyAndSum(int lastDig)
{
    int multiply = lastDig * 2;
    int i = 0;
    int lastDigMult;

    while (multiply > 0)
    {
        lastDigMult = multiply % 10;
        i = i + lastDigMult;
        multiply = multiply / 10;
    }
    return (i);
}

int size(long num)
{
    int i = 0;
    while (num > 0)
    {
        i++;
        num = num / 10;
    }
    return (i);
}

File: Python/credit.py
from cs50 import get_int


def main():
    while True:
        ccNumber = get_int("Card Number: ")
        if ccNumber >= 0:
            break

    if check_sum(ccNumber):
        print_brand(ccNumber)
    else:
        print("INVALID")


def check_sum(ccNumber):
    sum = 0
    digitCount = 0

    while ccNumber > 0:
        if digitCount % 2 == 0:
            sum += ccNumber % 10
        else:
            digit = 2 * (ccNumber % 10)
            sum += digit // 10 + digit % 10
        ccNumber //= 10
        digitCount += 1

    return (sum % 10 == 0)


def print_brand(ccNumber):

    ccNumber_str = str(ccNumber)

    if ccNumber_str.startswith(("34", "37")) and len(ccNumber_str) == 15:
        print("AMEX")
    elif ccNumber_str.startswith(("51", "52", "53", "54", "55")) and len(ccNumber_str) == 16:
        print("MASTERCARD")
    elif ccNumber_str.startswith("4") and (len(ccNumber_str) == 13 or len(ccNumber_str) == 16) and int(ccNumber_str[0]) == 4:
        print("VISA")
    else:
        print("INVALID")


if __name__ == "__main__":
    main()
